### DISCUSSION TITLE
ESP32 +8720 Ethernet module+ EWebServer - linker error - multiple definitions

### DISCUSSION THREAD

**Discussion by sergshadrap** [khoih-prog/EthernetWebServer#68] [Q&A] [answered]
Hi sir, could you help me figure my mistake out?
Just trying to move some quite big project has built for esp8266 to esp32 Ethernet , decided to try your library for this purpose, installed everything as described, checked with  examples - everything ok. Trying to compile my code in Arduino (1.8.19) -the compiler passes successfully, the linker show multiple errors , like:

```
c:/users/serg/documents/arduinodata/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2-patch3/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: sketch\Mpp32Parameters.cpp.o: in function `EthernetClient::write(unsigned char)':
C:\Users\serg\Documents\Arduino\libraries\Ethernet_Generic-main\src/EthernetClient_Impl.h:164: multiple definition of `EthernetClient::write(unsigned char)'; sketch\Mpp32Device.cpp.o:C:\Users\serg\Documents\Arduino\libraries\Ethernet_Generic-main\src/EthernetClient_Impl.h:164: first defined here
c:/users/serg/documents/arduinodata/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2-patch3/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: sketch\Mpp32Parameters.cpp.o: in function `EthernetClient::operator==(EthernetClient const&)':
C:\Users\serg\Documents\Arduino\libraries\Ethernet_Generic-main\src/EthernetClient_Impl.h:306: multiple definition of `EthernetClient::operator==(EthernetClient const&)'; sketch\Mpp32Device.cpp.o:C:\Users\serg\Documents\Arduino\libraries\Ethernet_Generic-main\src/EthernetClient_Impl.h:306: first defined here
c:/users/serg/documents/arduinodata/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2-patch3/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: sketch\Mpp32Parameters.cpp.o: in function `EthernetServer::write(unsigned char)':
```

and etc, and etc....
From the list of the errors I see that two library crosses each other : Ethernet_Generic and yours - how can I handle the problem?
Thanks in advance.

**Comment by khoih-prog** [comment]
Hi @sergshadrap 

Without an [**MRE**](https://stackoverflow.com/help/minimal-reproducible-example), I'm afraid nobody can guess what's wrong with your code. Especially you didn't specify which Ethernet and how you're using the libraries

I suggest you start from the simple ESP32 example, such as [**AdvancedWebServer_ESP32_SPI2**](https://github.com/khoih-prog/EthernetWebServer/tree/master/examples/AdvancedWebServer_ESP32_SPI2) to be sure everything is OK at your end (settings, versions, etc.), then moving on gradually, small step-by-step.

Good Luck,

**Reply by sergshadrap** [reply]
Thank you for your attention!
To be honest I thought , I exposed all the information might be needed. Sorry for that. Once again - I use ESP32 Wroom-E module with 8720 Lan Ethernet chip( as in the header). Using your library "EthernetWebServer". Accordingly to the advices you give in the readme, I've installed  Ethernet_Generic library. My usage of the library starts with the EthernetWeb server class where I want to substitute ESP8266WebServer.h library :

```cpp
#include <Arduino.h>
#include <EthernetWebServer.h>

#ifndef MPPPARAMETERS_H_
#define MPPPARAMETERS_H_

class MppParameters {
public:
	MppParameters(EthernetWebServer* webServer);
	String getParameter(const char *name);
	bool hasParameter(const char *name);
	String getParameterName(int i);
	String getParameter(int i);
	int getParameterCount();
	unsigned getUnsignedParameter(const char* name);
	float getFloatParameter(const char* name);
	int getIntParameter(const char* name);
	bool getBoolParameter(const char* name);
	String getAsQuery();
	ethernetHTTPUpload& getUpload() { return webRequest->upload(); }
protected:
private:
	EthernetWebServer* webRequest;
};

#endif /* MPPPARAMETERS_H_ */
```

I would share all the code with you , but there are 7 files with thousand strings of the code....
But initialization of Ethernet usage I  begined with Eth.h (Ethernet.h) library for ESP32.

```
#define ETH_CLK_MODE    ETH_CLOCK_GPIO0_OUT          // Version with PSRAM
#define ETH_POWER_PIN   4
#define ETH_TYPE        ETH_PHY_LAN8720
#define ETH_ADDR        0
#define ETH_MDC_PIN     23
#define ETH_MDIO_PIN    18
```

and further - 

```cpp
ETH.begin(ETH_ADDR, ETH_POWER_PIN, ETH_MDC_PIN, ETH_MDIO_PIN, ETH_TYPE, ETH_CLK_MODE)
```

Once again- All your samples has being compiled successfully. 
 I'd want to say that there are no any compiler errors - the compiler passes OK. My question is about the problem multiple definitions of the same functions,  I suspect that I can't use Eth.h and EthernetWebServer.h together?

**Comment by khoih-prog** [comment, marked as answer]
That's why I'm so confused. Sorry.

You don't need this library for **ESP32 + LAN8720**, such as **WT32_ETH01**

Use my [**WebServer_WT32_ETH01** library](https://github.com/khoih-prog/WebServer_WT32_ETH01) with built-in ESP32's  [**WebServer**](https://github.com/espressif/arduino-esp32/tree/master/libraries/WebServer).

As specified in [**Currently supported Ethernet shields/modules**](https://github.com/khoih-prog/EthernetWebServer#currently-supported-ethernet-shieldsmodules), this library is only supporting W5x00, ENC28J60, Teensy 4.1 QNEthernet and Portenta_H7 Ethernet.

Good Luck,

**Reply by sergshadrap** [reply]
Ou! Thanks ! That works. BTW in different examples you've got different class name , that is a mistake or they are aliased ?
in WebServer.ino you wrote "WiFiServer server(80);" , but in AdvancedWebServer -"WebServer server(80);"

**Reply by khoih-prog** [reply]
They are totally different classes.

**WiFiServer**, even with more sophisticated name, is simpler with simple features, defined in [**WiFiServer.h**](https://github.com/espressif/arduino-esp32/blob/master/libraries/WiFi/src/WiFiServer.h), provides lower level functions, such as

https://github.com/espressif/arduino-esp32/blob/c1dc4f2e24609049af261123e269a69accdba087/libraries/WebServer/src/WebServer.h#L122-L125

```cpp
void send(int code, const char* content_type = NULL, const String& content = String(""));
void send(int code, char* content_type, const String& content);
void send(int code, const String& content_type, const String& content);
void send(int code, const char* content_type, const char* content);
```

While **WebServer**, defined in [**WebServer.h**](https://github.com/espressif/arduino-esp32/blob/master/libraries/WebServer/src/WebServer.h), provides more complex `Web Server` features, such as

https://github.com/khoih-prog/WebServer_WT32_ETH01/blob/811a7d27cb07998817eececd7dacdc16b2dbfc10/examples/AdvancedWebServer/AdvancedWebServer.ino#L161-L169

```cpp
  server.on(F("/"), handleRoot);
  server.on(F("/test.svg"), drawGraph);
  server.on(F("/inline"), []()
  {
    server.send(200, F("text/plain"), F("This works as well"));
  });


  server.onNotFound(handleNotFound);
  server.begin();
```

**Reply by sergshadrap** [reply]
Ok. Got it. Thanks again.
